File: java-mvc/layered-architecture/src/main/java/com/ecommerce/layered/presentation/controller/ProductController.java
package com.ecommerce.layered.presentation.controller;

import com.ecommerce.layered.business.service.ProductService;
import com.ecommerce.layered.presentation.dto.CreateProductRequest;
import com.ecommerce.layered.presentation.dto.ProductResponse;
import jakarta.validation.Valid;
import org.springframework.data.domain.Page;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.math.BigDecimal;
import java.util.List;
import java.util.UUID;

/**
 * PRESENTATION LAYER - REST Controller.
 * Handles HTTP requests and delegates to Business layer.
 * Dependencies flow DOWNWARD: Presentation → Business → Data.
 */
@RestController
@RequestMapping("/api/v1/products")
public class ProductController {

    private final ProductService productService;

    public ProductController(ProductService productService) {
        this.productService = productService;
    }

    @PostMapping
    public ResponseEntity<ProductResponse> createProduct(@Valid @RequestBody CreateProductRequest request) {
        var entity = productService.createProduct(
                request.name(), request.description(), request.price(),
                request.category(), request.stockQuantity(), request.sku()
        );
        return ResponseEntity.status(HttpStatus.CREATED).body(ProductResponse.fromEntity(entity));
    }

    @GetMapping("/{id}")
    public ResponseEntity<ProductResponse> getProduct(@PathVariable UUID id) {
        var entity = productService.getProductById(id);
        return ResponseEntity.ok(ProductResponse.fromEntity(entity));
    }

    @GetMapping
    public ResponseEntity<Page<ProductResponse>> getAllProducts(
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "20") int size) {
        Page<ProductResponse> products = productService.getAllProducts(page, size)
                .map(ProductResponse::fromEntity);
        return ResponseEntity.ok(products);
    }

    @GetMapping("/search")
    public ResponseEntity<List<ProductResponse>> searchProducts(
            @RequestParam(required = false) String query,
            @RequestParam(required = false) String category,
            @RequestParam(required = false) BigDecimal minPrice,
            @RequestParam(required = false) BigDecimal maxPrice) {
        List<ProductResponse> products = productService.searchProducts(query, category, minPrice, maxPrice)
                .stream()
                .map(ProductResponse::fromEntity)
                .toList();
        return ResponseEntity.ok(products);
    }

    @PutMapping("/{id}")
    public ResponseEntity<ProductResponse> updateProduct(
            @PathVariable UUID id,
            @Valid @RequestBody CreateProductRequest request) {
        var entity = productService.updateProduct(id, request.name(),
                request.description(), request.price(), request.category());
        return ResponseEntity.ok(ProductResponse.fromEntity(entity));
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<Void> deleteProduct(@PathVariable UUID id) {
        productService.deleteProduct(id);
        return ResponseEntity.noContent().build();
    }

    @PatchMapping("/{id}/stock/decrease")
    public ResponseEntity<Void> decreaseStock(@PathVariable UUID id, @RequestParam int quantity) {
        productService.decreaseStock(id, quantity);
        return ResponseEntity.ok().build();
    }

    @PatchMapping("/{id}/stock/increase")
    public ResponseEntity<Void> increaseStock(@PathVariable UUID id, @RequestParam int quantity) {
        productService.increaseStock(id, quantity);
        return ResponseEntity.ok().build();
    }
}

File: dotnet-mvc/LayeredArchitecture/Presentation/Controllers/ProductController.cs
using LayeredArchitecture.Business.Services;
using LayeredArchitecture.Data.Entities;
using LayeredArchitecture.Presentation.Dto;
using Microsoft.AspNetCore.Mvc;

namespace LayeredArchitecture.Presentation.Controllers;

/// <summary>
/// REST controller for Product operations.
/// This belongs to the PRESENTATION LAYER — the topmost layer in the layered architecture.
/// It depends directly on the Business layer (ProductService).
/// Dependencies flow downward: Presentation → Business → Data.
/// </summary>
[ApiController]
[Route("api/v1/products")]
[Produces("application/json")]
public class ProductController : ControllerBase
{
    private readonly ProductService _productService;

    public ProductController(ProductService productService)
    {
        _productService = productService;
    }

    /// <summary>
    /// Creates a new product.
    /// </summary>
    [HttpPost]
    [ProducesResponseType(typeof(ProductResponse), StatusCodes.Status201Created)]
    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status400BadRequest)]
    public async Task<IActionResult> CreateProduct([FromBody] CreateProductRequest request)
    {
        var entity = await _productService.CreateProduct(
            request.Name, request.Description ?? string.Empty, request.Price,
            request.Category, request.StockQuantity, request.Sku);

        var response = MapToResponse(entity);
        return CreatedAtAction(nameof(GetProductById), new { id = entity.Id }, response);
    }

    /// <summary>
    /// Lists products with pagination.
    /// </summary>
    [HttpGet]
    [ProducesResponseType(typeof(PaginatedResponse<ProductResponse>), StatusCodes.Status200OK)]
    public async Task<IActionResult> ListProducts([FromQuery] int page = 0, [FromQuery] int pageSize = 20)
    {
        var (items, totalCount) = await _productService.ListProducts(page, pageSize);
        var effectivePageSize = Math.Min(Math.Max(pageSize, 1), 100);
        var totalPages = (int)Math.Ceiling((double)totalCount / effectivePageSize);

        var response = new PaginatedResponse<ProductResponse>(
            items.Select(MapToResponse),
            page,
            effectivePageSize,
            totalCount,
            totalPages
        );

        return Ok(response);
    }

    /// <summary>
    /// Gets a product by its ID.
    /// </summary>
    [HttpGet("{id:guid}")]
    [ProducesResponseType(typeof(ProductResponse), StatusCodes.Status200OK)]
    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status404NotFound)]
    public async Task<IActionResult> GetProductById(Guid id)
    {
        var entity = await _productService.GetProductById(id);
        return Ok(MapToResponse(entity));
    }

    /// <summary>
    /// Searches products by name, category, or description.
    /// </summary>
    [HttpGet("search")]
    [ProducesResponseType(typeof(IEnumerable<ProductResponse>), StatusCodes.Status200OK)]
    public async Task<IActionResult> SearchProducts([FromQuery] string query = "")
    {
        var entities = await _productService.SearchProducts(query);
        return Ok(entities.Select(MapToResponse));
    }

    /// <summary>
    /// Updates an existing product.
    /// </summary>
    [HttpPut("{id:guid}")]
    [ProducesResponseType(typeof(ProductResponse), StatusCodes.Status200OK)]
    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status404NotFound)]
    public async Task<IActionResult> UpdateProduct(Guid id, [FromBody] UpdateProductRequest request)
    {
        var entity = await _productService.UpdateProduct(
            id, request.Name, request.Description ?? string.Empty,
            request.Price, request.Category, request.Sku);

        return Ok(MapToResponse(entity));
    }

    /// <summary>
    /// Soft-deletes a product (sets active=false).
    /// </summary>
    [HttpDelete("{id:guid}")]
    [ProducesResponseType(StatusCodes.Status204NoContent)]
    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status404NotFound)]
    public async Task<IActionResult> DeleteProduct(Guid id)
    {
        await _productService.DeleteProduct(id);
        return NoContent();
    }

    /// <summary>
    /// Decreases product stock by the specified quantity.
    /// </summary>
    [HttpPatch("{id:guid}/stock/decrease")]
    [ProducesResponseType(typeof(ProductResponse), StatusCodes.Status200OK)]
    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status400BadRequest)]
    public async Task<IActionResult> DecreaseStock(Guid id, [FromQuery] int quantity)
    {
        var entity = await _productService.DecreaseStock(id, quantity);
        return Ok(MapToResponse(entity));
    }

    /// <summary>
    /// Increases product stock by the specified quantity.
    /// </summary>
    [HttpPatch("{id:guid}/stock/increase")]
    [ProducesResponseType(typeof(ProductResponse), StatusCodes.Status200OK)]
    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status400BadRequest)]
    public async Task<IActionResult> IncreaseStock(Guid id, [FromQuery] int quantity)
    {
        var entity = await _productService.IncreaseStock(id, quantity);
        return Ok(MapToResponse(entity));
    }

    private static ProductResponse MapToResponse(ProductEntity entity)
    {
        return new ProductResponse(
            entity.Id,
            entity.Name,
            entity.Description,
            entity.Price,
            entity.Category,
            entity.StockQuantity,
            entity.Sku,
            entity.Active,
            entity.CreatedAt,
            entity.UpdatedAt
        );
    }
}
